### PR TITLE
fix(stack-client): Get Icon Url uses preloaded url when oAuth not needed

### PR DESCRIPTION
**Pull request by trollepierre**
Need upgrade in cozy-ui as well + test + working without breaking change

Capture d’écran 2022-02-25 à 16 21 44

https://github.com/cozy/cozy-ui/pull/2064

### CODE REVIEW

File: packages/cozy-stack-client/src/getIconURL.preloader.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      const loadError = new Error(`Error while preloading ${source}`)
+      console.error(loadError.message)
+      reject(loadError)
+    }
+    loader.src = source
+    console.log('loader end, all set:')
+    console.log({loader})
+
+  })
+}
```
Comment (Crash--) [inline packages/cozy-stack-client/src/getIconURL.preloader.js:47]: not sure that this `preload/ed` stuff is needed. I don't see the point to have it now. Cozy-Stack should send the correct header to put it in cache, no?

Comment (trollepierre) [reply packages/cozy-stack-client/src/getIconURL.preloader.js:47]: exact, I'm updating the PR right now.

File: packages/cozy-stack-client/src/getIconURL.spec.js [hunk 2]
```diff
@@ -39,142 +41,192 @@ describe('get icon', () => {
...
+      stackClient.uri = undefined
+      const url = await getIconURL(stackClient, defaultOpts)
+      expect(url).toEqual(new ErrorReturned())
+    })
 
-    expect(global.URL.createObjectURL).not.toHaveBeenCalled()
+    it('should catch occurring error and returns nothing - ErrorReturned', async () => {
+      defaultOpts.appData = { latest_version: undefined, slug: 'slug' }
+      const url = await getIconURL(stackClient, defaultOpts)
+      expect(url).toEqual(new ErrorReturned())
```
Comment (trollepierre) [inline packages/cozy-stack-client/src/getIconURL.spec.js:229]: @Crash-- : I wonder if this catching error mechanism is wanted

File: packages/cozy-stack-client/src/getIconURL.spec.js [hunk 3]
```diff
@@ -39,142 +41,192 @@ describe('get icon', () => {
...
+    it('should return url from registry if app data not containing icon path', async () => {
+      defaultOpts.appData = { latest_version: { version: '3' }, slug: 'slug' }
+      const url = await getIconURL(stackClient, defaultOpts)
+      expect(url).toEqual('http://cozy.tools:8080/registry/slug/3/icon')
+    })
+
+    it('should catch occurring error and returns nothing - ErrorReturned', async () => {
+      stackClient.uri = undefined
+      const url = await getIconURL(stackClient, defaultOpts)
+      expect(url).toEqual(new ErrorReturned())
```
Comment (trollepierre) [inline packages/cozy-stack-client/src/getIconURL.spec.js:223]: @Crash-- : I wonder if this catching error mechanism is wanted

File: packages/cozy-stack-client/src/getIconURL.spec.js [hunk 4]
```diff
@@ -39,142 +41,192 @@ describe('get icon', () => {
...
-      })
-    )
-    expect(url.data[0]).toBe('<svg></svg>')
-    resetcreateObjectURL()
-    await getIconUrlDefault(stackClient, defaultOpts)
+    it('should return url from appData if appData is a string', async () => {
+      defaultOpts.appData = 'appData-that-is-a-string'
+      const url = await getIconURL(stackClient, defaultOpts)
+      expect(url).toEqual(
+        'http://cozy.tools:8080/registry/appData-that-is-a-string/icon'
```
Comment (trollepierre) [inline packages/cozy-stack-client/src/getIconURL.spec.js:210]: @Crash-- : I wonder if this appData that can be string is possible

Comment (Crash--) [reply packages/cozy-stack-client/src/getIconURL.spec.js:210]: Nein. I don't think so.

File: packages/cozy-stack-client/src/getIconURL.spec.js (hunk 3 above)
Comment (Crash--) [reply packages/cozy-stack-client/src/getIconURL.spec.js:223]: it's wanted

File: packages/cozy-stack-client/src/getIconURL.spec.js (hunk 2 above)
Comment (Crash--) [reply packages/cozy-stack-client/src/getIconURL.spec.js:229]: wanted

File: packages/cozy-stack-client/src/getIconURL.spec.js (hunk 4 above)
Comment (trollepierre) [reply packages/cozy-stack-client/src/getIconURL.spec.js]: this code + test have been removed

**Review by Crash-- — APPROVED** [review]

**Pull request merged**